File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    
    col_length = len(maps[0])
    row_length = len(maps)
    visited = [[0 for col in range(col_length)] for row in range(row_length)]
    
    queue = deque([])
    queue.append((0,0))
    visited[0][0] = 1
    
    while queue:
        x, y = queue.popleft()
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if nx < 0 or ny < 0 or nx >= row_length or ny >= col_length:
                continue
                
            if maps[nx][ny] == 0:
                continue
                
            if visited[nx][ny] == 0:
                queue.append((nx,ny))
                visited[nx][ny] += visited[x][y] + 1

    if visited[-1][-1] == 0:
        return -1
    else:
        return visited[-1][-1]

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[] dx = {-1,1,0,0};
    int[] dy = {0,0,-1,1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        int[][] visited = new int[maps.length][maps[0].length];
        bfs(maps,visited);
        answer = visited[maps.length-1][maps[0].length-1];
        
        if(answer == 0)
            return -1;
        else
            return answer;
    }
    
    public void bfs(int[][] maps, int[][] visited) {
        int x = 0;
        int y = 0;
        
        Queue<int[]> queue = new LinkedList<>();
        visited[x][y] = 1;
        queue.add(new int[]{x,y});
        
        while(!queue.isEmpty()) {
            int[] current = queue.remove();
            
            for(int i=0; i < 4; i++) {
                int nx = current[0] + dx[i];
                int ny = current[1] + dy[i];
                
                if(nx < 0 || ny < 0 || nx >= maps.length || ny >= maps[0].length)
                    continue;
                
                if(visited[nx][ny] == 0 && maps[nx][ny] == 1) {
                    visited[nx][ny] = visited[current[0]][current[1]] + 1;
                    queue.add(new int[]{nx,ny});
                }
            }
        }
    }
}